### PR TITLE
Update to laminas-coding-standard 2.0

### PR DESCRIPTION
**Pull request by arueckauer**
|    Q          |   A
|-------------- | ------
| Documentation | yes
| BC Break      | yes
| QA            | yes

### Description

This PR updates `laminas-coding-standard` to version `2.0` and provides necessary code changes.

Incompatible parts are being ignored now and need to be revisited in a later change. Most, if not all, incompatibilities are because DocBlocks instead of type declarations are used for reflection.

### CODE REVIEW

File: composer.json
Comment (weierophinney) [inline composer.json]: Please use `~2.0.0`. New minor versions may introduce new sniffs... which would make CS checks fail. By using the `~` constraint, we can opt-in to those new sniffs when we are ready, instead of having to suddenly update for a new patch.

File: docs/book/v3/reference/migration.md
Comment (weierophinney) [inline docs/book/v3/reference/migration.md]: We should probably introduce the `Interface` suffix variants in a new minor release on the 2.X series, and have the old, non-suffixed versions extend them. This will allow deprecation.

Ideally, we should also provide a laminas-cli command in this package to help users update any implementations they have to the new variants; that should be done in the 2.x series as well.

File: src/AbstractServer.php
Comment (weierophinney) [inline src/AbstractServer.php]: Follow the guidelines in the docblock here, and you can then get rid of the `phpcs:disable`/`phpcs:enable` dance. :wink: (make sure you also update any calls to the method)

Honestly, I'd likely add `private function buildCallback()` in v2, and have `_buildCallback()` proxy to it. Since it's an internal method, we can do this safely, and `_buildCallback()` (a) could be marked `@deprecated`, and (b) would never be called anyways, making overriding it by end-users moot. Then for 3.0, we just remove the underscore-prefixed version.

I'd do that for all underscore-prefixed methods except `_buildSignature()` and `_dispatch()` (which are the only methods extensions will consume).

That would, of course, mean having a new minor version before we merge this particular PR. But I think that it's a worthwhile exercise.

File: src/AbstractServer.php
Comment (weierophinney) [inline src/AbstractServer.php]: I'd do two things with this method:

- Remove the underscore prefix (per the docblock!)
- Make the method `final` (so it cannot be overridden by end-users)

This will allow getting rid of the phpcs dance, and also fulfill the ideas already sketched for 3.0.

**Review by weierophinney — CHANGES_REQUESTED** [review]
I think there are a few things we need to do in a new minor release before finalizing this patch.

File: composer.json
Comment (arueckauer) [reply composer.json]: 👍🏻 Thanks for the explanation. Done.

**Comment by weierophinney** [comment]
@arueckauer I've merged the changes from #8 to the dev-3.0 branch; please rebase and push again. You should be able to also remove the deprecated items on this commit! :tada:

File: docs/book/v3/reference/migration.md
Comment (arueckauer) [reply docs/book/v3/reference/migration.md]: The new interfaces will be introduced in 2.9.

For the command to work, one would execute the command, updating userland code to the usage of the new interfaces, right? That would mean, a minor update to make the command available and immediately upgrade to 3.x, because the new interfaces are not accepted in 2.x as arguments. To me it would make more sense to upgrade directly to 3.x and execute the command. Maybe I am missing the value of an intermediate minor update(?).

File: src/AbstractServer.php
Comment (arueckauer) [reply src/AbstractServer.php]: If I haven't missed anything, everything mentioned here should be covered. Let me know if further changes are required. Resolving this for now.

**Comment by arueckauer** [comment]
@weierophinney Rebase is done and changes are provided. Ready for your review. Only pendent point is the laminas-cli command, which needs clarification (could be provided outside of this PR). Thank you.

File: docs/book/v3/reference/migration.md
Comment (weierophinney) [reply docs/book/v3/reference/migration.md]: I got it backwards: have the _new_ interfaces extend the _old_interfaces.

We did this in laminas-eventmanager and laminas-servicemanager going into v3. By introducing them in v2 releases, users were able to update their code to implement the new interfaces, while keeping backwards compatibility. Then, when they upgraded to version 3, they did not need to make any further changes.

Most of those consumers are Laminas packages. This essentially allows them to pin to **either** the 2.X series **or** the 3.X series, if they update their interfaces.

Now that we have a stable release of laminas-cli (stable in that there is a version released that you can pin to semantically in order to keep a stable API — we dropped version 0.1.0 this morning), you can even go one step further: provide a tool to help the user migrate their implementations to the new interfaces. Something like `laminas server:migrate:v3-interfaces .`, which would go through the current directory, find implementations, and update them.

Of course: all of this is moot if the v2 and v3 interfaces are not compatible. If that's the case, then scratch the whole thing, and instead just provide a command in v3 for helping migrate to the new interfaces (or have us supply [Rector configuration](https://getrector.org/) for doing that).

File: docs/book/v3/reference/migration.md
Comment (weierophinney) [inline docs/book/v3/reference/migration.md]: I wouldn't mention that this is a change dictated by laminas-coding-standard; I'd mention it's being done for consistency with the rest of the project, where we have standardized on the `Interface` suffix.

File: src/AbstractServer.php
Comment (weierophinney) [inline src/AbstractServer.php]:
```suggestion
            return $function(...$params);
```

We're bumping the PHP version. We can use the fact that there is now uniform callable syntax and the splat operator. :)

File: src/AbstractServer.php
Comment (weierophinney) [inline src/AbstractServer.php]:
```suggestion
            $callback = [$class, $method];
            return $callback(...$params);
```

Same comment here; simplify! Doing so removes the need for ignoring CS rules.

File: src/AbstractServer.php
Comment (weierophinney) [inline src/AbstractServer.php]:
```suggestion
        $callback = [$object, $method];
        return $callback(...$params);
```

Same comment here.

File: src/AbstractServer.php
Comment (weierophinney) [inline src/AbstractServer.php:140]: We might want to document the _purpose_ of this method, for those implementing it.

File: src/ClientInterface.php
Comment (weierophinney) [inline src/ClientInterface.php]: Why are you removing the parameter typehints here?

File: src/Reflection/Node.php
Comment (weierophinney) [inline src/Reflection/Node.php:46]: Why are you having phpcs ignore this block?

File: src/Reflection/Node.php
Comment (weierophinney) [inline src/Reflection/Node.php:100]: Docblock says `return mixed`, but declaration says `void`...

File: test/Reflection/ReflectionFunctionTest.php
Comment (weierophinney) [inline test/Reflection/ReflectionFunctionTest.php:206]: These are largely a remnant from the days when visibility wasn't used often. I think it might make more sense to update these, add visibility, and update the docblocks, instead of ignoring phpcs.

File: test/TestAsset/ReflectionTestClass.php
Comment (weierophinney) [inline test/TestAsset/ReflectionTestClass.php]: Same comment here: update the asset to fix CS, versus ignoring CS.

File: test/TestAsset/reflectionTestFunction.php
Comment (weierophinney) [inline test/TestAsset/reflectionTestFunction.php:27]: And here, too.

**Review by weierophinney — CHANGES_REQUESTED** [review]
Getting closer!

In a previous comment, I've indicated what I think should happen with regards to interface changes and tooling; I'll leave the final decision to you on that.

I've flagged some additional improvements to make, mostly providing ways we can NOT ignore CS, and instead fix the code so it doesn't fail CS.

File: docs/book/v3/reference/migration.md
Comment (arueckauer) [reply docs/book/v3/reference/migration.md]: > I got it backwards: have the new interfaces extend the _old_interfaces.

That would have been nice to know in #8.

File: src/AbstractServer.php
Comment (arueckauer) [reply src/AbstractServer.php]: Edit: Change provided to replace `call_user_func_array`.

File: src/AbstractServer.php
Comment (arueckauer) [reply src/AbstractServer.php]: Edit: Change provided to replace `call_user_func_array`.

File: src/AbstractServer.php
Comment (arueckauer) [reply src/AbstractServer.php]: Edit: Change provided to replace `call_user_func_array`.

File: src/AbstractServer.php
Comment (arueckauer) [reply src/AbstractServer.php:140]: Agreed. However, this is a documentation issue which is IMO out-of-scope of this PR. I opened an issue for that.

File: src/Reflection/Node.php
Comment (arueckauer) [reply src/Reflection/Node.php:46]: CS validation will fail because of the FQCN (expecting `Node`). The DocBlock param type is currently used in reflection. Change it to `Node` will fail test `ReflectionMethodTest::testMethodDocBlockFromParent`. With argument types in place, I would like to check whether it is possible to drop the DocBlock reflection in favor of using the argument types instead. I will have to do some research to see, if it will work (problematic could be mixed types).

File: src/Reflection/Node.php
Comment (arueckauer) [reply src/Reflection/Node.php:100]: The argument is hinted with `mixed`. Return type hint in DocBlock is obsolete due to return type declaration.

File: test/Reflection/ReflectionFunctionTest.php
Comment (arueckauer) [reply test/Reflection/ReflectionFunctionTest.php:206]: CS is ignored here, since DocBlock is required for reflection of the component. CS will fail validation due to redundant param and return tags. Removing DocBlock will fail test `ReflectionFunctionTest::testMultipleWhitespaceBetweenDoctagsAndTypes`.

File: test/TestAsset/reflectionTestFunction.php
Comment (arueckauer) [reply test/TestAsset/reflectionTestFunction.php:27]: In the current implementation where DocBlock param tags are parsed, it is not possible to satisfy CS. This has to be reviewed with a possible change of parsing argument type declarations.

File: docs/book/v3/reference/migration.md
Comment (arueckauer) [reply docs/book/v3/reference/migration.md]: Done.

File: src/ClientInterface.php
Comment (arueckauer) [reply src/ClientInterface.php]: You are right, this makes no sense. I am not sure, what introduced this. Re-adding type declarations.

File: test/TestAsset/ReflectionTestClass.php
Comment (arueckauer) [reply test/TestAsset/ReflectionTestClass.php]: Change provided.

File: docs/book/v3/reference/migration.md
Comment (arueckauer) [reply docs/book/v3/reference/migration.md]: PR with changes for 2.9 is provided. I opened an issue for a possible migration command (see #11).

**Comment by arueckauer** [comment]
@weierophinney Thanks for the feedback. I provided the requested changes.

There are three discussions left, all addressing the same CS issue caused by the component's usage of DocBlocks. As mentioned, I would like to investigate an improved solution by using type declarations and dropping support for DocBlocks altogether. This can be addressed after completion of the component maintenance: adding static analysis, bump to PHP 7.4 (or at least 7.3), add typed properties and introduce callables and Iterators. All these I would like to address in separate PRs.

File: src/Reflection/Node.php
Comment (weierophinney) [reply src/Reflection/Node.php:46]: Ah, right - forgot that the component had not yet been updated to get type information via Reflection instead of the docblocks!

I think that will likely work really well. For mixed or union types, you still need annotations, so we can very likely fallback to parsing annotations only if the typehint is missing.

But agreed - that's for another set of changes.

**Review by weierophinney — APPROVED** [review]
:ship:

I can't remember — did I give you write access to this repo yet, @arueckauer ? If not, I can merge, but if so, go for it!

**Comment by arueckauer** [comment]
@weierophinney You offered me becoming a maintainer, but we did not follow. For now I am happy, if you would merge it. Thank you.

**Comment by weierophinney** [comment]
Merged - I'll get you setup as a maintainer shortly!

**Pull request merged**
